sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1818,7 +1818,10 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
             "sqlExpr" -> a.sql,
             "cols" -> cols))
       }
-      resolved
+      resolved match {
```

Comment:
Without this change, nested references in MERGE are resolved as `t.struct_col.id AS id`, which breaks our assumption about what kind of keys we expect in assignments. It seems safe for both keys and values but let me know if there are any cases when it is not safe.